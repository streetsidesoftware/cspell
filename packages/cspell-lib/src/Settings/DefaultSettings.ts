import type {
    CSpellSettings,
    CSpellSettingsWithSourceTrace,
    PredefinedPatterns,
    RegExpPatternDefinition,
} from '@cspell/cspell-types';
import { resolveFile } from '../util/resolveFile';
import { readSettings } from './CSpellSettingsServer';
import { mergeSettings } from './index';
import * as LanguageSettings from './LanguageSettings';
import * as RegPat from './RegExpPatterns';

const defaultConfigFileModuleRef = '@cspell/cspell-bundled-dicts/cspell-default.json';

// Do not use require.resolve because webpack will mess it up.
const defaultConfigFile = resolveConfigModule(defaultConfigFileModuleRef);

const regExpSpellCheckerDisable = [
    RegPat.regExSpellingGuardBlock,
    RegPat.regExSpellingGuardLine,
    RegPat.regExSpellingGuardNext,
];

// cspell:ignore filetypes
const predefinedPatterns = [
    // Exclude patterns
    { name: 'CommitHash', pattern: RegPat.regExCommitHash },
    { name: 'CommitHashLink', pattern: RegPat.regExCommitHashLink },
    { name: 'CStyleHexValue', pattern: RegPat.regExCStyleHexValue },
    { name: 'CSSHexValue', pattern: RegPat.regExCSSHexValue },
    { name: 'Urls', pattern: RegPat.regExMatchUrls },
    { name: 'HexValues', pattern: RegPat.regExMatchCommonHexFormats },
    { name: 'SpellCheckerDisable', pattern: regExpSpellCheckerDisable },
    { name: 'PublicKey', pattern: RegPat.regExPublicKey },
    { name: 'RsaCert', pattern: RegPat.regExCert },
    { name: 'SshRsa', pattern: RegPat.regExSshRSA },
    { name: 'EscapeCharacters', pattern: RegPat.regExEscapeCharacters },
    { name: 'Base64', pattern: RegPat.regExBase64 },
    { name: 'Base64SingleLine', pattern: RegPat.regExBase64SingleLine },
    { name: 'Base64MultiLine', pattern: RegPat.regExBase64MultiLine },
    { name: 'Email', pattern: RegPat.regExEmail },
    { name: 'SHA', pattern: RegPat.regExSha },
    { name: 'HashStrings', pattern: RegPat.regExHashStrings },
    { name: 'UnicodeRef', pattern: RegPat.regExUnicodeRef },
    { name: 'UUID', pattern: RegPat.regExUUID },
    { name: 'href', pattern: RegPat.regExHRef },
    { name: 'SpellCheckerDisableBlock', pattern: RegPat.regExSpellingGuardBlock },
    { name: 'SpellCheckerDisableLine', pattern: RegPat.regExSpellingGuardLine },
    { name: 'SpellCheckerDisableNext', pattern: RegPat.regExSpellingGuardNext },
    { name: 'SpellCheckerIgnoreInDocSetting', pattern: RegPat.regExIgnoreSpellingDirectives },

    // Include Patterns
    { name: 'PhpHereDoc', pattern: RegPat.regExPhpHereDoc },
    { name: 'string', pattern: RegPat.regExString },
    { name: 'CStyleComment', pattern: RegPat.regExCStyleComments },
    { name: 'Everything', pattern: '.*' },
] as const;

type NameType<T> = T extends readonly { name: infer U }[] ? U : never;

type ExtendsType<T, U> = T extends U ? T : never;

type PredefinedPatternNames = ExtendsType<NameType<typeof predefinedPatterns>, PredefinedPatterns>;

const defaultRegExpPatterns: RegExpPatternDefinition[] = [...predefinedPatterns];

const definedDefaultRegExpExcludeList: PredefinedPatterns[] = [
    'SpellCheckerDisable',
    'SpellCheckerIgnoreInDocSetting',
    'Urls',
    'Email',
    'RsaCert',
    'SshRsa',
    'Base64MultiLine',
    'Base64SingleLine',
    'CommitHash',
    'CommitHashLink',
    'CStyleHexValue',
    'CSSHexValue',
    'SHA',
    'HashStrings',
    'UnicodeRef',
    'UUID',
];

// This bit of copying is done to have the complier ensure that the defaults exist.
const defaultRegExpExcludeList: PredefinedPatternNames[] = definedDefaultRegExpExcludeList;

export const _defaultSettings: Readonly<CSpellSettingsWithSourceTrace> = Object.freeze({
    id: 'static_defaults',
    language: 'en',
    name: 'Static Defaults',
    enabled: true,
    enabledLanguageIds: [
        'ada',
        'csharp',
        'go',
        'javascript',
        'javascriptreact',
        'json',
        'markdown',
        'mdx',
        'php',
        'plaintext',
        'python',
        'text',
        'typescript',
        'typescriptreact',
        'haskell',
        'html',
        'css',
        'less',
        'scss',
        'latex',
        'ruby',
        'rust',
        'shellscript',
        'toml',
    ],
    maxNumberOfProblems: 100,
    numSuggestions: 10,
    suggestionsTimeout: 500,
    suggestionNumChanges: 3,
    words: [],
    userWords: [],
    ignorePaths: [],
    allowCompoundWords: false,
    patterns: defaultRegExpPatterns,
    ignoreRegExpList: defaultRegExpExcludeList,
    languageSettings: LanguageSettings.getDefaultLanguageSettings(),
    source: { name: 'defaultSettings' },
    reporters: [],
});

const getSettings = (function () {
    let settings: CSpellSettings | undefined = undefined;
    return function () {
        if (!settings) {
            const jsonSettings = readSettings(defaultConfigFile);
            settings = mergeSettings(_defaultSettings, jsonSettings);
            settings.name = jsonSettings.name;
        }
        return settings;
    };
})();

function resolveConfigModule(configModuleName: string) {
    return resolveFile(configModuleName, __dirname).filename;
}

export function getDefaultSettings(): CSpellSettings {
    return getSettings();
}
