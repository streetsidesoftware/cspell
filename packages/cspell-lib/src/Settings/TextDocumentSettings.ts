import type { CSpellSettings, CSpellUserSettings } from '@cspell/cspell-types';
import { CSpellSettingsInternal } from '../Models/CSpellSettingsInternalDef';
import * as CSpellSettingsServer from './CSpellSettingsServer';
import { getInDocumentSettings } from './InDocSettings';
import { calcSettingsForLanguageId } from './LanguageSettings';

export function combineTextAndLanguageSettings(
    settings: CSpellUserSettings,
    text: string,
    languageId: string | string[]
): CSpellSettingsInternal {
    const docSettings = extractSettingsFromText(text);
    const settingsForText = CSpellSettingsServer.mergeSettings(settings, docSettings);
    const langSettings = calcSettingsForLanguageId(settingsForText, languageId);
    // Merge again, to force In-Doc settings.
    return CSpellSettingsServer.mergeSettings(langSettings, docSettings);
}

export function extractSettingsFromText(text: string): CSpellSettings {
    return getInDocumentSettings(text);
}
