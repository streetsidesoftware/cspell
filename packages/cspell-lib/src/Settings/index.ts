export * from './CSpellSettingsServer';
export * from './DefaultSettings';
export * from '@cspell/cspell-types';
export * from './ImportError';
