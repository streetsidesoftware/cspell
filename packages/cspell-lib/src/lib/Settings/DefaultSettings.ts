import type { PredefinedPatterns, RegExpPatternDefinition } from '@cspell/cspell-types';
import { parsers } from 'cspell-grammar';

import { srcDirectory } from '../../lib-cjs/index.cjs';
import type { CSpellSettingsInternal } from '../Models/CSpellSettingsInternalDef.js';
import { createCSpellSettingsInternal } from '../Models/CSpellSettingsInternalDef.js';
import { PatternRegExp } from '../Models/PatternRegExp.js';
import { resolveFile } from '../util/resolveFile.js';
import { readSettings } from './Controller/configLoader/index.js';
import { mergeSettings } from './CSpellSettingsServer.js';
import * as LanguageSettings from './LanguageSettings.js';
import * as RegPat from './RegExpPatterns.js';

const defaultConfigFileModuleRef = '@cspell/cspell-bundled-dicts/cspell-default.json';

// Do not use require.resolve because webpack will mess it up.
const defaultConfigFile = resolveConfigModule(defaultConfigFileModuleRef);

const regExpSpellCheckerDisable = [
    new PatternRegExp(RegPat.regExSpellingGuardBlock),
    new PatternRegExp(RegPat.regExSpellingGuardLine),
    new PatternRegExp(RegPat.regExSpellingGuardNext),
];

// cspell:ignore filetypes
const predefinedPatterns = [
    // Exclude patterns
    { name: 'CommitHash', pattern: RegPat.regExCommitHash },
    { name: 'CommitHashLink', pattern: RegPat.regExCommitHashLink },
    { name: 'CStyleHexValue', pattern: RegPat.regExCStyleHexValue },
    { name: 'CSSHexValue', pattern: RegPat.regExCSSHexValue },
    { name: 'Urls', pattern: RegPat.regExMatchUrls },
    { name: 'HexValues', pattern: RegPat.regExMatchCommonHexFormats },
    { name: 'SpellCheckerDisable', pattern: regExpSpellCheckerDisable },
    { name: 'PublicKey', pattern: RegPat.regExPublicKey },
    { name: 'RsaCert', pattern: RegPat.regExCert },
    { name: 'SshRsa', pattern: RegPat.regExSshRSA },
    { name: 'EscapeCharacters', pattern: RegPat.regExEscapeCharacters },
    { name: 'Base64', pattern: RegPat.regExBase64 },
    { name: 'Base64SingleLine', pattern: RegPat.regExBase64SingleLine },
    { name: 'Base64MultiLine', pattern: RegPat.regExBase64MultiLine },
    { name: 'Email', pattern: RegPat.regExEmail },
    { name: 'SHA', pattern: RegPat.regExSha },
    { name: 'HashStrings', pattern: RegPat.regExHashStrings },
    { name: 'UnicodeRef', pattern: RegPat.regExUnicodeRef },
    { name: 'UUID', pattern: RegPat.regExUUID },
    { name: 'href', pattern: RegPat.regExHRef },
    { name: 'SpellCheckerDisableBlock', pattern: RegPat.regExSpellingGuardBlock },
    { name: 'SpellCheckerDisableLine', pattern: RegPat.regExSpellingGuardLine },
    { name: 'SpellCheckerDisableNext', pattern: RegPat.regExSpellingGuardNext },
    { name: 'SpellCheckerIgnoreInDocSetting', pattern: RegPat.regExIgnoreSpellingDirectives },

    // Include Patterns
    { name: 'PhpHereDoc', pattern: RegPat.regExPhpHereDoc },
    { name: 'string', pattern: RegPat.regExString },
    { name: 'CStyleComment', pattern: RegPat.regExCStyleComments },
    { name: 'Everything', pattern: '.*' },
] as const;

type NameType<T> = T extends readonly { name: infer U }[] ? U : never;

type ExtendsType<T, U> = T extends U ? T : never;

type PredefinedPatternNames = ExtendsType<NameType<typeof predefinedPatterns>, PredefinedPatterns>;

const defaultRegExpPatterns: RegExpPatternDefinition[] = [...predefinedPatterns].map(normalizePattern);

const definedDefaultRegExpExcludeList: PredefinedPatterns[] = [
    'SpellCheckerDisable',
    'SpellCheckerIgnoreInDocSetting',
    'Urls',
    'Email',
    'RsaCert',
    'SshRsa',
    'Base64MultiLine',
    'Base64SingleLine',
    'CommitHash',
    'CommitHashLink',
    'CStyleHexValue',
    'CSSHexValue',
    'SHA',
    'HashStrings',
    'UnicodeRef',
    'UUID',
];

// This bit of copying is done to have the complier ensure that the defaults exist.
const defaultRegExpExcludeList: PredefinedPatternNames[] = definedDefaultRegExpExcludeList;

export const _defaultSettingsBasis: Readonly<CSpellSettingsInternal> = Object.freeze(
    createCSpellSettingsInternal({
        id: 'static_defaults',
        language: 'en',
        name: 'Static Defaults',
        enabled: true,
        enabledLanguageIds: [],
        maxNumberOfProblems: 100,
        numSuggestions: 10,
        suggestionsTimeout: 500,
        suggestionNumChanges: 3,
        words: [],
        userWords: [],
        ignorePaths: [],
        allowCompoundWords: false,
        patterns: defaultRegExpPatterns,
        ignoreRegExpList: [],
        languageSettings: [],
        source: { name: 'defaultSettings' },
        reporters: [],
        plugins: [{ parsers }],
    }),
);

export const _defaultSettings: Readonly<CSpellSettingsInternal> = Object.freeze(
    createCSpellSettingsInternal({
        ..._defaultSettingsBasis,
        enabledLanguageIds: [
            'ada',
            'csharp',
            'go',
            'javascript',
            'javascriptreact',
            'json',
            'markdown',
            'mdx',
            'php',
            'plaintext',
            'python',
            'text',
            'typescript',
            'typescriptreact',
            'haskell',
            'html',
            'css',
            'less',
            'scss',
            'latex',
            'ruby',
            'rust',
            'shellscript',
            'toml',
        ],
        ignoreRegExpList: defaultRegExpExcludeList,
        languageSettings: LanguageSettings.getDefaultLanguageSettings(),
    }),
);

const getSettings = (function () {
    let settings: CSpellSettingsInternal | undefined = undefined;
    return function (useDefaultDictionaries: boolean) {
        if (!useDefaultDictionaries) {
            return _defaultSettingsBasis;
        }
        if (!settings) {
            const jsonSettings = readSettings(defaultConfigFile);
            settings = mergeSettings(_defaultSettings, jsonSettings);
            if (jsonSettings.name !== undefined) {
                settings.name = jsonSettings.name;
            } else {
                delete settings.name;
            }
        }
        return settings;
    };
})();

function resolveConfigModule(configModuleName: string) {
    return resolveFile(configModuleName, srcDirectory).filename;
}

function normalizePattern(pat: RegExpPatternDefinition): RegExpPatternDefinition {
    const { name, pattern, description } = pat;
    if (!(pattern instanceof RegExp)) return pat;

    return {
        name,
        pattern: new PatternRegExp(pattern),
        description,
    };
}

export function getDefaultSettings(useDefaultDictionaries = true): CSpellSettingsInternal {
    return getSettings(useDefaultDictionaries);
}

export function getDefaultBundledSettings(): CSpellSettingsInternal {
    return getDefaultSettings();
}
