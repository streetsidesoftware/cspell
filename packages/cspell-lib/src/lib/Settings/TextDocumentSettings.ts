import type { CSpellSettings, CSpellUserSettings } from '@cspell/cspell-types';

import type { CSpellSettingsInternal } from '../Models/CSpellSettingsInternalDef.js';
import { mergeSettings, toInternalSettings } from './CSpellSettingsServer.js';
import { getInDocumentSettings } from './InDocSettings.js';
import { calcSettingsForLanguageId } from './LanguageSettings.js';

export function combineTextAndLanguageSettings(
    settings: CSpellUserSettings,
    text: string | undefined,
    languageId: string | string[],
): CSpellSettingsInternal {
    if (!text) {
        return toInternalSettings(calcSettingsForLanguageId(settings, languageId));
    }
    const docSettings = extractSettingsFromText(text);
    const settingsForText = mergeSettings(settings, docSettings);
    const langSettings = calcSettingsForLanguageId(settingsForText, languageId);
    // Merge again, to force In-Doc settings.
    return mergeSettings(langSettings, docSettings);
}

export function extractSettingsFromText(text: string): CSpellSettings {
    return getInDocumentSettings(text);
}
