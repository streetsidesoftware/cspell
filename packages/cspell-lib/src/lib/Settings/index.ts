export { currentSettingsFileVersion, ENV_CSPELL_GLOB_ROOT } from './constants';
export {
    clearCachedSettingsFiles,
    defaultConfigFilenames,
    defaultFileName,
    extractImportErrors,
    getCachedFileSize,
    getGlobalSettings,
    loadConfig,
    loadPnP,
    loadPnPSync,
    readRawSettings,
    readSettings,
    readSettingsFiles,
    searchForConfig,
    sectionCSpell,
} from './Controller/configLoader';
export { ImportError } from './Controller/ImportError';
export type { ConfigurationDependencies, ImportFileRefWithError } from './CSpellSettingsServer';
export {
    calcOverrideSettings,
    checkFilenameMatchesGlob,
    extractDependencies,
    finalizeSettings,
    getSources,
    mergeInDocSettings,
    mergeSettings,
} from './CSpellSettingsServer';
export { getDefaultBundledSettings, getDefaultSettings } from './DefaultSettings';
