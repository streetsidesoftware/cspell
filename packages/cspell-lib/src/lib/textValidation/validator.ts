import type { CSpellUserSettings } from '@cspell/cspell-types';
import { IssueType } from '@cspell/cspell-types';

import { createTextDocument } from '../Models/TextDocument.js';
import type { ValidationIssue } from '../Models/ValidationIssue.js';
import * as Settings from '../Settings/index.js';
import type { DirectiveIssue } from '../Settings/InDocSettings.js';
import { validateInDocumentSettings } from '../Settings/InDocSettings.js';
import { CompoundWordsMethod, getDictionaryInternal } from '../SpellingDictionary/index.js';
import { settingsToValidateOptions } from './settingsToValidateOptions.js';
import { validateText as validateFullText } from './textValidator.js';
import type { ValidateTextOptions } from './ValidateTextOptions.js';

export const diagSource = 'cSpell Checker';

/**
 * @deprecated
 * @deprecationMessage Use spellCheckDocument
 */
export async function validateText(
    text: string,
    settings: CSpellUserSettings,
    options: ValidateTextOptions = {}
): Promise<ValidationIssue[]> {
    const finalSettings = Settings.finalizeSettings(settings);
    const dict = await getDictionaryInternal(finalSettings);
    const spellingIssues = [...validateFullText(text, dict, settingsToValidateOptions(finalSettings))];
    const validationIssues =
        options.validateDirectives || finalSettings.validateDirectives
            ? validateInDocumentSettings(text, settings)
            : [];
    const issues: ValidationIssue[] = spellingIssues.concat(mapValidationIssues(text, validationIssues));
    if (!options.generateSuggestions) {
        return issues;
    }
    const sugOptions = {
        numSuggestions: options.numSuggestions,
        compoundMethod: CompoundWordsMethod.NONE,
        includeTies: false,
        ignoreCase: !(settings.caseSensitive ?? false),
        timeout: settings.suggestionsTimeout,
        numChanges: settings.suggestionNumChanges,
    };
    const withSugs = issues.map((t) => {
        const text = t.text;
        const suggestionsEx = dict
            .suggest(text, sugOptions)
            .map(({ word, isPreferred }) => (isPreferred ? { word, isPreferred } : { word }));
        t.suggestions = suggestionsEx.map((s) => s.word);
        t.suggestionsEx = suggestionsEx;
        return t;
    });

    return withSugs;
}

function mapValidationIssues(text: string, valIssues: Iterable<DirectiveIssue>): ValidationIssue[] {
    const issues = [...valIssues];
    if (!issues.length) return [];

    const document = createTextDocument({ uri: '', content: text });
    const issueType = IssueType.directive;

    function toValidationIssue(dirIssue: DirectiveIssue): ValidationIssue {
        const { text, range, suggestions, suggestionsEx, message } = dirIssue;
        const offset = range[0];
        const pos = document.positionAt(offset);
        const line = document.getLine(pos.line);
        const issue: ValidationIssue = { text, offset, line, suggestions, suggestionsEx, message, issueType };

        return issue;
    }

    return issues.map(toValidationIssue);
}
