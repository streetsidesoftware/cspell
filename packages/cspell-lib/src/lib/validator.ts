export type {
    CheckTextInfo,
    IncludeExcludeOptions,
    TextInfoItem,
    ValidateTextOptions,
    ValidationIssue,
} from './textValidation/index.js';
export { checkText, checkTextDocument, IncludeExcludeFlag, validateText } from './textValidation/index.js';
