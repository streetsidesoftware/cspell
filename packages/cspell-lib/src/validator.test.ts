import type { CSpellSettings } from '@cspell/cspell-types';
import { loremIpsum } from 'lorem-ipsum';
import { getDefaultSettings } from './Settings/DefaultSettings';
import * as tds from './Settings/TextDocumentSettings';
import * as Validator from './validator';
import { IncludeExcludeFlag } from './validator';

// cSpell:ignore brouwn jumpped lazzy wrongg mispelled ctrip nmove mischecked

describe('Validator', () => {
    test('validates the validator', async () => {
        const text = 'The quick brouwn fox jumpped over the lazzy dog.';
        const languageId = 'plaintext';
        const settings = getSettings(text, languageId);
        const results = await Validator.validateText(text, settings);
        const words = results.map(({ text }) => text);
        expect(words).toEqual(['brouwn', 'jumpped', 'lazzy']);
    });

    test('validates ignore Case', async () => {
        const text = 'The Quick brown fox Jumped over the lazy dog.';
        const languageId = 'plaintext';
        const settings = getSettings(text, languageId);
        const results = await Validator.validateText(text, settings);
        const words = results.map(({ text }) => text);
        expect(words).toEqual([]);
    });

    test('validate limit', async () => {
        const text = loremIpsum({ count: 5, units: 'paragraphs' });
        const languageId = 'plaintext';
        const settings = { ...getSettings(text, languageId), maxNumberOfProblems: 10 };
        const results = await Validator.validateText(text, settings);
        expect(results).toHaveLength(10);
    });

    test('validates reserved words', async () => {
        const text = 'constructor const prototype type typeof null undefined';
        const languageId = 'javascript';
        const settings = { ...getSettings(text, languageId), maxNumberOfProblems: 10 };
        const results = await Validator.validateText(text, settings);
        expect(results).toHaveLength(0);
    });

    test('validates regex inclusions/exclusions', async () => {
        const text = sampleCode;
        const languageId = 'plaintext';
        const settings = { ...getSettings(text, languageId), maxNumberOfProblems: 10 };
        const results = await Validator.validateText(text, settings);
        const words = results.map((wo) => wo.text);
        expect(words).toEqual(expect.arrayContaining(['wrongg']));
        expect(words).toEqual(expect.arrayContaining(['mispelled']));
        expect(words).toEqual(expect.not.arrayContaining(['xaccd']));
        expect(words).toEqual(expect.not.arrayContaining(['ctrip']));
        expect(words).toEqual(expect.not.arrayContaining(['FFEE']));
        expect(words).toEqual(expect.not.arrayContaining(['nmove']));
    });

    test('validates ignoreRegExpList', async () => {
        const text = sampleCode;
        const languageId = 'plaintext';
        const settings = {
            ...getSettings(text, languageId),
            maxNumberOfProblems: 10,
            ignoreRegExpList: ['^const [wy]RON[g]+', 'mis.*led'],
        };
        const results = await Validator.validateText(text, settings);
        const words = results.map((wo) => wo.text);
        expect(words).toEqual(expect.not.arrayContaining(['wrongg']));
        expect(words).toEqual(expect.not.arrayContaining(['mispelled']));
        expect(words).toEqual(expect.arrayContaining(['mischecked']));
    });

    test('validates ignoreRegExpList 2', async () => {
        const results = await Validator.validateText(sampleCode, {
            ignoreRegExpList: ['/^const [wy]ron[g]+/gim', '/MIS...LED/g', '/mischecked'],
        });
        const words = results.map((wo) => wo.text);
        expect(words).toEqual(expect.not.arrayContaining(['wrongg']));
        expect(words).toEqual(expect.arrayContaining(['mispelled']));
        expect(words).toEqual(expect.arrayContaining(['mischecked']));
    });

    test('validates malformed ignoreRegExpList', async () => {
        const results = await Validator.validateText(sampleCode, {
            ignoreRegExpList: ['/wrong[/gim', 'mis.*led'],
        });
        const words = results.map((wo) => wo.text);
        expect(words).toEqual(expect.arrayContaining(['wrongg']));
        expect(words).toEqual(expect.not.arrayContaining(['mispelled']));
        expect(words).toEqual(expect.arrayContaining(['mischecked']));
    });

    // cspell:ignore hellosd applesq bananasa respectss
    test('Issue #7', async () => {
        const text = `Fails to detect obviously misspelt words, such as:
            hellosd
            applesq
            bananasa
            respectss
        `;
        const expected = ['hellosd', 'applesq', 'bananasa', 'respectss'];
        const languageId = 'plaintext';
        const settings = getSettings(text, languageId);
        const results = await Validator.validateText(text, settings);
        const words = results.map(({ text }) => text);
        expect(words.sort()).toEqual(expected.sort());
    });

    test('Validates contractions', async () => {
        const text = `
            We have a bit of text to check. Don't look too hard.
            Which single quote to use? Is it shouldn't or shouldnâ€™t?
        `;
        const languageId = 'plaintext';
        const settings = getSettings(text, languageId);
        const results = await Validator.validateText(text, settings);
        const words = results.map(({ text }) => text);
        expect(words.sort()).toEqual([]);
    });
    test('tests calcIncludeExcludeInfo', async () => {
        const words = sampleWords;
        const info = await Validator.checkText(sampleText, { words, ignoreRegExpList: [/The/g] });
        const strings = info.items.map((a) => a.text);
        expect(strings).toHaveLength(17);
        expect(strings.join('')).toBe(sampleText);

        let last = 0;
        info.items.forEach((i) => {
            expect(i.startPos).toBe(last);
            last = i.endPos;
        });
        expect(last).toBe(sampleText.length);
    });

    // cspell:ignore witth feww mistaks
    test('validateText with suggestions', async () => {
        const text = `
            Here is a bit of text witth a feww mistaks.
        `;
        const languageId = 'plaintext';
        const settings = getSettings(text, languageId);
        const result = await Validator.validateText(text, settings, { generateSuggestions: true, numSuggestions: 5 });
        expect(result).toMatchSnapshot();
    });

    test('tests calcIncludeExcludeInfo exclude everything', async () => {
        const words = sampleWords;
        const info = await Validator.checkText(sampleText, {
            words,
            ignoreRegExpList: [/(.|\s)+/],
        });
        const result = info.items.map((a) => a.text);
        expect(result).toHaveLength(1);
        expect(result.join('')).toBe(sampleText);
        expect(info.items[0].flagIE).toBe(IncludeExcludeFlag.EXCLUDE);
    });

    test('tests calcIncludeExcludeInfo include everything', async () => {
        const words = sampleWords;
        const info = await Validator.checkText(sampleText, { words });
        const result = info.items.map((a) => a.text);
        expect(result).toHaveLength(9);
        expect(result.join('')).toBe(sampleText);
        expect(info.items[0].flagIE).toBe(IncludeExcludeFlag.INCLUDE);
    });

    // const isFoundTrue = { isFound: true };
    const isFoundFalse = { isFound: false };
    const isFlaggedTrue = { isFlagged: true };
    const isFlaggedFalse = { isFlagged: false };

    // cspell:ignore grappes
    test.each`
        text              | expected
        ${'hello'}        | ${[]}
        ${'flagged'}      | ${[mValIssue('flagged', isFlaggedTrue)]}
        ${'grappes'}      | ${[mValIssue('grappes', isFoundFalse, isFlaggedFalse)]}
        ${'colour'}       | ${[mValIssue('colour', isFlaggedTrue)]}
        ${'ignored'}      | ${[]}
        ${'crazzzy-code'} | ${[]}
        ${'hyphen-wordz'} | ${[]}
    `('validation "$text"', async ({ text, expected }) => {
        const settings = sampleSettings();
        const languageId = 'plaintext';
        const docSettings = tds.combineTextAndLanguageSettings(settings, text, languageId);
        const results = await Validator.validateText(text, docSettings);
        expect(results).toEqual(expected);
    });
});

// cspell:ignore xaccd ffee

const sampleCode = `

// Verify urls do not get checked.
const url = 'http://ctrip.com?q=words';

// Verify hex values.
const value = 0xaccd;

/* spell-checker:disable */

const weirdWords = ['ctrip', 'xebia', 'zando', 'zooloo'];

/* spell-checker:enable */

const wrongg = 'mispelled';
const check = 'mischecked';
const message = "\\nmove to next line";

const hex = 0xBADC0FFEE;

`;

// cspell:ignore lightbrown whiteberry redberry
const sampleText = `
    The elephant and giraffe
    The lightbrown worm ate the apple, mango, and, strawberry.
    The little ant ate the big purple grape.
    The orange tiger ate the whiteberry and the redberry.
`;

const sampleWords = [
    'and',
    'ant',
    'apple',
    'ate',
    'big',
    'elephant',
    'giraffe',
    'grape',
    'little',
    'mango',
    'orange',
    'purple',
    'the',
    'tiger',
    'worm',
];

const flagWords = ['hte', 'flagged', 'ignored'];

const rejectWords = ['!colour', '!behaviour', '!favour'];

// cspell:ignore crazzzy wordz
const ignoreWords = ['ignored', 'crazzzy-code'];

const words = sampleWords.concat(rejectWords).concat(['hyphen-wordz']);

const sampleCSpell: CSpellSettings = {
    ...getDefaultSettings(),
    version: '0.2',
    flagWords,
    words,
    ignoreWords,
};

const defaultSettings: CSpellSettings = {
    ...getDefaultSettings(),
    enabledLanguageIds: ['plaintext', 'javascript'],
};

function mValIssue(text: string, ...parts: Partial<Validator.ValidationIssue>[]): Validator.ValidationIssue {
    const issue: Partial<Validator.ValidationIssue> = {
        text,
    };
    for (const p of parts) {
        Object.assign(issue, p);
    }
    return oc<Validator.ValidationIssue>(issue);
}

function sampleSettings() {
    return Object.freeze({ ...sampleCSpell });
}

function getSettings(text: string, languageId: string) {
    return tds.combineTextAndLanguageSettings(defaultSettings, text, languageId);
}

function oc<T>(t: Partial<T>): T {
    return expect.objectContaining(t);
}
