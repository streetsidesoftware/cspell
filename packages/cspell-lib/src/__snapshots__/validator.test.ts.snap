// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Validator validateText with suggestions 1`] = `
Array [
  Object {
    "isFlagged": false,
    "isFound": false,
    "line": Object {
      "offset": 1,
      "text": "            Here is a bit of text witth a feww mistaks.
",
    },
    "offset": 35,
    "suggestions": Array [
      "witty",
      "witt",
      "witch",
      "with",
      "Witt",
      "width",
    ],
    "text": "witth",
  },
  Object {
    "isFlagged": false,
    "isFound": false,
    "line": Object {
      "offset": 1,
      "text": "            Here is a bit of text witth a feww mistaks.
",
    },
    "offset": 43,
    "suggestions": Array [
      "few",
      "fewer",
      "few's",
      "fe's",
      "feal",
      "fear",
      "feat",
      "feck",
      "feds",
      "feed",
      "feel",
      "feer",
      "fees",
      "feet",
      "fell",
      "felt",
      "feme",
      "fend",
      "fens",
      "fern",
      "fess",
      "fest",
      "feta",
      "fete",
      "feud",
      "flews",
    ],
    "text": "feww",
  },
  Object {
    "isFlagged": false,
    "isFound": false,
    "line": Object {
      "offset": 1,
      "text": "            Here is a bit of text witth a feww mistaks.
",
    },
    "offset": 48,
    "suggestions": Array [
      "mistake",
      "mistakes",
      "mistaken",
      "mistaker",
      "mistake's",
      "mistakers",
    ],
    "text": "mistaks",
  },
]
`;
